File: NanoRpc.Net/src/RpcChannel.cs
namespace NanoRpc
{
    /// <summary>
    /// </summary>
    public abstract class RpcChannel
    {
        private readonly RpcController _rpcController;

        /// <summary>
        /// </summary>
        /// <param name="rpcController">
        /// </param>
        public RpcChannel(RpcController rpcController)
        {
            _rpcController = rpcController;
            _rpcController.Channel = this;
        }

        /// <summary>
        /// </summary>
        public abstract void Start();

        /// <summary>
        /// </summary>
        /// <param name="rpcMessage">
        /// </param>
        internal abstract void Send(RpcMessage rpcMessage);

        /// <summary>
        /// </summary>
        /// <param name="message">
        /// </param>
        protected void Receive(RpcMessage message)
        {
            _rpcController.Receive(message);
        }
    }
}

File: NanoRpc/src/rpc_channel.cpp

#include "rpc_channel.hpp"
#include "rpc_controller.hpp"

namespace NanoRpc {

RpcChannel::RpcChannel(RpcController *controller) : controller_(controller) {
  controller_->set_channel(this);
}

RpcChannel::~RpcChannel() {
  if (controller_ != NULL)
    controller_->set_channel(NULL);
  controller_ = NULL;
}

void RpcChannel::Receive(const RpcMessage &message) {
  controller_->Receive(message);
}

} // namespace
